### PR TITLE
Add replay option with user specified window location

### PR DESCRIPTION
**Pull request by mizhen**
**The problem**
   
Allow user to specified window orgin/starting location and replay GFXR trace in windowed mode without resizing the window resolution (it should be the same size as captured).

**The solution**

Add replay option with user specified window location. The feature is for both D3D12 and Vulkan on Windows platform. It allows user to specify window starting location and force playback in windowed mode with original resolution (instead of full screen).

In addition to the new option, the pull request also includes a bug fix to the following issue:
    --fw option failed to force windowed for full screen resolution. For example, for 4k trace file on 4k screen, the force-windowed option always fails if set --fw 3840,2160, it will always be full-screen mode.
	
**Result**
Tested DX12/Vulkan samples and target title trace file with the build of the pull request, all passed.

### CODE REVIEW

**Comment by perim** [comment]
How will you deal with multiple windows?  For the patrace project we record the window positions of each created window so that they can be recreated in the same position on replay. This is especially useful for UI apps that are composed of multiple windows side by side of each other. Could that be a solution here as well?

**Comment by mizhen** [comment]
> How will you deal with multiple windows? For the patrace project we record the window positions of each created window so that they can be recreated in the same position on replay. This is especially useful for UI apps that are composed of multiple windows side by side of each other. Could that be a solution here as well?

In current code (with or without the pull request), for supporting the multiple windows which are part of UI apps, current code need some changes such as recording the order, location, size and related style parameters of windows into some new trace block to replace current ResizeWindowCommand or ResizeWindowCommand2 for the window creation and resize during playback time, plus some other related changes. However, I think multiple windows of UI apps handling can be a new feature to implement in future.


As for this pull-request replaces existing fixed window origin 0,0 with user specified location and other related full screen / windowed mode changes, it doesn't append any handling with multiple windows of which the location has different values.

File: framework/decode/vulkan_virtual_swapchain.cpp [hunk 1]
```diff
@@ -608,8 +608,9 @@ VkResult VulkanVirtualSwapchain::AcquireNextImageKHR(VkResult                  o
     }
 
     result = func(device, swapchain, timeout, semaphore, fence, image_index);
-    if (result != VK_SUCCESS)
+    if ((result != VK_SUCCESS) && (result != VK_SUBOPTIMAL_KHR))
```
Comment (bradgrantham-lunarg) [inline framework/decode/vulkan_virtual_swapchain.cpp:611]: I don't think we want to start failing for this result code.  `SUBOPTIMAL` isn't a failure.  If you'd like, you could put in a TODO comment referencing https://github.com/LunarG/gfxreconstruct/issues/759 but take out these tests of `result`.

**Review by bradgrantham-lunarg — CHANGES_REQUESTED** [review]

File: framework/decode/vulkan_virtual_swapchain.cpp (hunk 1 above)
Comment (mizhen) [reply framework/decode/vulkan_virtual_swapchain.cpp:611]: Yes, VK_SUBOPTIMAL_KHR isn't a failure.

If the test is only (result != VK_SUCCESS),  VK_SUBOPTIMAL_KHR value will match it and proceed the following error output:
        GFXRECON_LOG_ERROR("Virtual swapchain failed AcquireNextImageKHR 0x%08x for swapchain (ID = %" PRIu64 ")",
                           result,
                           swapchain_info->capture_id);
                           
That's why the test (result != VK_SUBOPTIMAL_KHR) is added.

Comment (bradgrantham-lunarg) [reply framework/decode/vulkan_virtual_swapchain.cpp:611]: Oh!  I read this change as the opposite of what you've actually added.  Yes, this is correct.

**Comment by locke-lunarg** [comment]
Could we combine `force_windowed` and `force_windowed_origin` into one option? or do we prefer to keep them separating?

**Comment by mizhen** [comment]
> LGTM. But, could we combine `force_windowed` and `force_windowed_origin`? or do we prefer to separate?

Currently the function of the two options is different, one is --force-windowed <width,height>, another is --force-windowed-origin <xloc,yloc>. We can combine the two into --force-windowed <width,height,x,y> in the future when virtual swapchain supports both dx12 and vulkan and can scale to specific window size.

**Comment by locke-lunarg** [comment]
I think it's ok to combine them in the future. Please rebase it, and pass the test, and then it will be good to merge.

**Comment by mizhen** [comment]
Will fix the macOS build issues.

**Comment by mizhen** [comment]
> > I think it's ok to combine them in the future. Please rebase it, and pass the test, and then it will be good to merge.
> 
> OK, will do that.

Done.

**Review by locke-lunarg — APPROVED** [review]

**Pull request merged**